src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java

```diff
@@ -1266,27 +1273,98 @@ private void mergeFirstPartyDataFromUser(ExtUser userExt, ObjectNode result) {
+        final Map<Integer, List<Segment>> segments = CollectionUtils.emptyIfNull(data).stream()
                 .filter(Objects::nonNull)
-                .flatMap(segments -> segments.stream()
-                        .map(Segment::getId))
+                .map(RubiconBidder::getValidSegments)
                 .filter(Objects::nonNull)
-                .toList();
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        Map.Entry::getValue,
+                        (value1, value2) -> Stream.concat(value1.stream(), value2.stream())
+                                .collect(Collectors.toCollection(ArrayList::new))));
+
+        if (MapUtils.isEmpty(segments)) {
+            return;
+        }
+
+        final Map<Integer, List<String>> relevantSegments = pickRelevantSegments(segments);
+        final Map<String, List<String>> resultSegments = groupBySegtaxValues(relevantSegments, segtaxValues);
+
+        resultSegments.forEach((segtaxValue, segmentIds) -> {
+            final ArrayNode array = target.putArray(segtaxValue);
+            segmentIds.forEach(array::add);
+        });
+    }
+
+    private static Map<String, List<String>> groupBySegtaxValues(Map<Integer, List<String>> segments,
+                                                                  Set<Integer> segtaxValues) {
+
+        return segments.entrySet().stream()
+                .map(entry -> Map.entry(resolveSegmentName(entry.getKey(), segtaxValues), entry.getValue()))
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        Map.Entry::getValue,
+                        (value1, value2) -> Stream.concat(value1.stream(), value2.stream())
+                                .collect(Collectors.toCollection(ArrayList::new))));
+    }
+
+    private static String resolveSegmentName(Integer taxonomyId, Set<Integer> segtaxValues) {
+        return segtaxValues.contains(taxonomyId) ? SEGTAX_IAB : SEGTAX_TAX + taxonomyId;
+    }
 
-        if (CollectionUtils.isNotEmpty(iabValue)) {
-            final ArrayNode iab = target.putArray("iab");
-            iabValue.forEach(iab::add);
+    private static Map.Entry<Integer, List<Segment>> getValidSegments(Data data) {
+        final ObjectNode ext = data.getExt();
+        final JsonNode taxonomyId = ext != null ? ext.get(SEGTAX) : null;
+
+        if (taxonomyId != null && taxonomyId.isInt()) {
+            final List<Segment> segments = getValidOnlySegments(data.getSegment());
+            if (CollectionUtils.isNotEmpty(segments)) {
+                return Map.entry(taxonomyId.intValue(), segments);
+            }
+        }
+
+        return null;
+    }
+
+    private static List<Segment> getValidOnlySegments(List<Segment> segments) {
+        return segments.stream()
+                .filter(segment -> StringUtils.isNotBlank(segment.getId()))
+                .collect(Collectors.toCollection(ArrayList::new));
+    }
```

Comment:
Because, the main interest for us is the last elements, let's use `Deque` here:
```java
    private static Deque<Segment> getValidOnlySegments(List<Segment> segments) {
        return CollectionUtils.isNotEmpty(segments)
                ? segments.stream()
                .filter(segment -> StringUtils.isNotBlank(segment.getId()))
                .collect(Collectors.toCollection(ArrayDeque::new))
                : null;
    }
```